### PR TITLE
HHH-16759 - Merge fails when entity has an Embedded Java record

### PR DESCRIPTION
**Pull request by cigaly**
Jira issue [HHH-16759](https://hibernate.atlassian.net/browse/HHH-16759)

[HHH-16759]: https://hibernate.atlassian.net/browse/HHH-16759?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by cigaly** [comment]
Passing all JDK 17 test with H2 and PostgreSQL

**Comment by jrenaat** [comment]
Looks good, but there is a small type in the test's name, if you could correct that, please?

File: hibernate-core/src/main/java/org/hibernate/type/ComponentType.java [hunk 1]
```diff
@@ -623,7 +630,7 @@ public CascadeStyle getCascadeStyle(int i) {
 
 	@Override
 	public boolean isMutable() {
-		return compositeUserType == null || compositeUserType.isMutable();
+		return (compositeUserType == null || compositeUserType.isMutable()) && !ReflectHelper.isRecord( componentClass );
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/type/ComponentType.java:633]: IMO this should be stored in a field in `ComponentType`, as the `isRecord` check is not cheap.

Comment (cigaly) [reply hibernate-core/src/main/java/org/hibernate/type/ComponentType.java:633]: You are right, I will fix that.

**Comment by cigaly** [comment]
@jrenaat 
> Looks good, but there is a small type in the test's name, if you could correct that, please?

I am not sure that I see what you are talking about.

File: hibernate-core/src/test/java17/org/hibernate/orm/test/records/MergeRecordPropetyTestCase.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+import jakarta.persistence.Id;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+@DomainModel(annotatedClasses = {
+		MergeRecordPropetyTestCase.MyEntity.class
+})
+@SessionFactory
+@JiraKey("HHH-16759")
+public class MergeRecordPropetyTestCase {
```
Comment (jrenaat) [inline hibernate-core/src/test/java17/org/hibernate/orm/test/records/MergeRecordPropetyTestCase.java:21]: "MergeRecordPropetyTestCase" should be "MergeRecordPropertyTestCase" (it's one of those things that either you see it at once, or you will never see it - our brain plays this kind of tricks on us ;) )

Comment (cigaly) [reply hibernate-core/src/test/java17/org/hibernate/orm/test/records/MergeRecordPropetyTestCase.java:21]: I see ... time to check my eyesight :-(

File: hibernate-core/src/main/java/org/hibernate/type/ComponentType.java [hunk 3]
```diff
@@ -623,7 +632,7 @@ public CascadeStyle getCascadeStyle(int i) {
 
 	@Override
 	public boolean isMutable() {
-		return compositeUserType == null || compositeUserType.isMutable();
+		return componentClassIsNotRecord && ( compositeUserType == null || compositeUserType.isMutable() );
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/type/ComponentType.java:635]:
```suggestion
		return mutable;
```

File: hibernate-core/src/main/java/org/hibernate/type/ComponentType.java [hunk 4]
```diff
@@ -69,6 +71,7 @@ public ComponentType(Component component, int[] originalPropertyOrder, MetadataB
 		this.componentClass = component.isDynamic()
 				? Map.class
 				: component.getComponentClass();
+		this.componentClassIsNotRecord = !ReflectHelper.isRecord( componentClass );
```
[the remark this answers is not in the document]
Comment (cigaly) [reply hibernate-core/src/main/java/org/hibernate/type/ComponentType.java:74]: OK, but I will move this at the bottom of constructor, after compositeUserType is set.

**Pull request merged**